src/include/duckdb/optimizer/join_order/plan_enumerator.hpp

```diff
@@ -0,0 +1,89 @@
+//===----------------------------------------------------------------------===//
+//                         DuckDB
+//
+// duckdb/optimizer/join_order/plan_enumerator.hpp
+//
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "duckdb/common/unordered_map.hpp"
+#include "duckdb/common/unordered_set.hpp"
+#include "duckdb/optimizer/join_order/join_relation.hpp"
+#include "duckdb/optimizer/join_order/cardinality_estimator.hpp"
+#include "duckdb/optimizer/join_order/query_graph.hpp"
+#include "duckdb/optimizer/join_order/join_node.hpp"
+#include "duckdb/optimizer/join_order/cost_model.hpp"
+#include "duckdb/parser/expression_map.hpp"
+#include "duckdb/planner/logical_operator.hpp"
+#include "duckdb/planner/logical_operator_visitor.hpp"
+
+#include <functional>
+
+namespace duckdb {
+
+class PlanEnumerator {
+public:
+	explicit PlanEnumerator(QueryGraphManager &query_graph_manager, CostModel &cost_model,
+	                        const QueryGraphEdges &query_graph)
+	    : query_graph(query_graph), query_graph_manager(query_graph_manager), cost_model(cost_model),
+	      full_plan_found(false), must_update_full_plan(false) {
+	}
+
+	//! Perform the join order solving
+	unique_ptr<JoinNode> SolveJoinOrder(bool force_no_cross_product);
+	void InitLeafPlans();
+
+	static unique_ptr<LogicalOperator> BuildSideProbeSideSwaps(unique_ptr<LogicalOperator> plan);
+
+private:
+	QueryGraphEdges const &query_graph;
+	//! The total amount of join pairs that have been considered
+	idx_t pairs = 0;
+	//! The set of edges used in the join optimizer
+	QueryGraphManager &query_graph_manager;
+	//! Cost model to evaluate cost of joins
+	CostModel &cost_model;
+	//! A map to store the optimal join plan found for a specific JoinRelationSet*
+	unordered_map<JoinRelationSet *, unique_ptr<JoinNode>> plans;
```

Comment:
Can we maybe use a `reference_map_t` here? It's basically a map with a pointer as a key, but it fits more nicely into our memory safety standards.

I know this is how it currently is on master, would just be nice to get rid of the raw pointer here.